libs/native/src/main/java/org/elasticsearch/nativeaccess/Systemd.java

```diff
@@ -41,15 +45,56 @@ public void notify_stopping() {
     }
 
     private void notify(String state, boolean warnOnError) {
-        int rc = lib.sd_notify(0, state);
-        logger.trace("sd_notify({}, {}) returned [{}]", 0, state, rc);
-        if (rc < 0) {
-            String message = String.format(Locale.ROOT, "sd_notify(%d, %s) returned error [%d]", 0, state, rc);
-            if (warnOnError) {
-                logger.warn(message);
+        int sockfd = libc.socket(PosixCLibrary.AF_UNIX, PosixCLibrary.SOCK_DGRAM, 0);
+        if (sockfd < 0) {
+            throwOrLog("Could not open systemd socket: " + libc.strerror(libc.errno()), warnOnError);
+            return;
+        }
+        RuntimeException error = null;
+        try {
+            var sockAddr = libc.newUnixSockAddr(socketPath);
+            if (libc.connect(sockfd, sockAddr) != 0) {
+                throwOrLog("Could not connect to systemd socket: " + libc.strerror(libc.errno()), warnOnError);
+                return;
+            }
+            buffer.buffer().clear();
```

Comment:
I'll look into this. I thought the existing solution already wasn't threadsafe, but i see that we do possibly call these from different threads in SystemdPlugin

Reply:
Since this isn't performance sensitive (we only do a couple of these calls, usually not near each other), I just add a simple synchronized block on the buffer around its use. Happy to change that further.